build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1185,6 +1417,156 @@ message ExecuteRequest {
   ResultsCachePolicy results_cache_policy = 8;
 }
 
+// Metadata and requests to the client for an ongoing
+// [dynamic execution][build.bazel.remote.execution.v2.Execution.ExecuteWithDynamicRetrieval],
+// which will be contained in the [metadata field][google.longrunning.Operation.response] of the
+// [Operation][google.longrunning.Operation].
+//
+// Design note: Dynamic directory and file requests are made using individual
+// messages for each directory or file path. The typical implementation
+// utilizes a virtual filesystem and a directory or file read request is
+// held pending in the VFS awaiting the information from the client.
+// VFS rendering callbacks may occur in a multithreaded fashion, but this is
+// unusual, and a chatty back-and-forth request series to the client resolves
+// pending I/O faster for each thread rather than adding in a Nagle
+// algorithm implementation and using repeated fields for the dynamic requests.
+// Dynamic callbacks for content are expected to be unusual or rare by
+// construction, and inherently slow down action execution.
+message DynamicExecuteOperationMetadata {
+  // Used by the worker to request that the client upload specific content
+  // to the CAS.
+  message DynamicFileContentRequest {
+    // The file path for which the worker does not have the contents.
+    // Depending on the server and client implementation this may be a
+    // relative or fully qualified path.
+    string path = 1;
+
+    // The file hash of the content expected by the worker.
+    // If this is not the current hash of the file content at `path`
+    // then the corrected hash is returned in
+    // `DynamicExecuteOperationMetadata.dynamic_directory_metadata_response`.
+    string file_hash = 2;
```

Comment:
I'm a bit confused about why there's a hash here at all. Is there any reason that a worker would know what hash it expects a file to have, but that file is missing from the CAS, other than that it just made a request for the hash?

It seems to me that when a request is made for a file's hash, the client should also upload the file's contents so that they are available for the worker to use.

Reply:
Changed this to a Digest, and updated doc comments for the response with Eric's commentary.